src/dmd/dtoh.d

```diff
@@ -0,0 +1,1900 @@
+        environment.reset(7); // erase cached environment updates
+        arch = parse_arch_arg(&dflags, arch);
+    }
+
+    bool is64bit = arch[0] == '6';
+
+    version(Windows) // delete LIB entry in [Environment] (necessary for optlink) to allow inheriting environment for MS-COFF
+    if (is64bit || strcmp(arch, "32mscoff") == 0)
+        environment.update("LIB", 3).ptrvalue = null;
+
+    // read from DFLAGS in [Environment{arch}] section
+    char[80] envsection = void;
+    sprintf(envsection.ptr, "Environment%s", arch);
+    sections.push(envsection.ptr);
+    parseConfFile(&environment, global.inifilename, inifilepath, inifile.len, inifile.buffer, &sections);
+    getenv_setargv(readFromEnv(&environment, "DFLAGS"), &arguments);
+    updateRealEnvironment(&environment);
+    environment.reset(1); // don't need environment cache any more
+
+    if (parseCommandLine(arguments, args.length, global.params, files))
+    {
+        Loc loc;
+        errorSupplemental(loc, "run 'dmd -man' to open browser on manual");
+        return 1;
+    }
+
+
+    /*version(BUILD_COMPILER)
+    {
+        global.path.push(druntimeFullPath.toStringz());
+        //TODO: fixme for LDC
+        global.filePath.push(__FILE_FULL_PATH__.dirName.buildPath("../../").toStringz());
+        global.filePath.push(__FILE_FULL_PATH__.dirName.buildPath("../../res/").toStringz());
+    }*/
+
+    DMDType._init();
+    version(BUILD_COMPILER)
+    {
+        DMDModule._init();
+        DMDClass._init();
+    }
+
+    setVersions();
+
+    Modules modules;
+
+    string path = __FILE_FULL_PATH__.dirName.buildPath("../dmd/");
+    version (BUILD_COMPILER)
+        auto srcs = sources;
+    else
+        auto srcs = args [1 .. $];
+    foreach (f; srcs)
+    {
+        string fn = buildPath(path, f);
+
+        auto id = Identifier.idPool(baseName(fn, ".d"));
+        auto m = new Module(fn.toStringz(), id, false, false);
+        auto input = readText(fn);
+
+        if (!Module.rootModule)
+            Module.rootModule = m;
+
+        m.importedFrom = m;
+        m.srcfile.setbuffer(cast(void*)input.ptr, input.length);
+        m.srcfile._ref = 1;
+        m.parse();
+        modules.push(m);
+    }
+
+    foreach (m; modules)
+        m.importAll(null);
+    foreach (m; modules)
+        m.dsymbolSemantic(null);
+
+    Module.dprogress = 1;
+    Module.runDeferredSemantic();
+
+    foreach (m; modules)
+        m.semantic2(null);
+    Module.runDeferredSemantic2();
+
+    foreach (m; modules)
+        m.semantic3(null);
+    Module.runDeferredSemantic3();
+
+    OutBuffer buf;
+    genCppFiles(&buf, &modules);
+
+    printf("%s\n", buf.peekString());
+    return 0;
+}
+}
+
+void setVersions()
+{
+    import dmd.cond : VersionCondition;
+    version(BUILD_COMPILER)
+    {
+        VersionCondition.addPredefinedGlobalIdent("NoBackend");
+        VersionCondition.addPredefinedGlobalIdent("NoMain");
+    }
+}
+
+string dirName(string path)
```

Comment:
This already exists in `dmd/mars.d`. Should then probably be moved to util, s.t. it can be reused.